enhancements/cvo/clusteroperator-resource-handling.md

```diff
@@ -0,0 +1,100 @@
+---
+title: clusteroperator-resource-handling
+authors:
+  - "@deads2k"
+reviewers:
+  - "@abhinavdahiya"
+  - "@sdodson"
+approvers:
+  - "@abhinavdahiya"
+  - "@derekwaynecarr"
+creation-date: 2019-12-11
+last-updated: 2019-12-11
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# ClusterOperator Resource Handling
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+`clusteroperators.config.openshift.io` are used to determine success of installation.
+They are also used to drive collection of debugging data from tools like `oc adm inspect` and `oc adm must-gather`.
+The `clusteroperator` resources in a payload should always be present, even before installation of particular
+operators to see which `clusteroperators` need to report in and to allow establishing `.status.relatedResources` before
+an operator pod runs.
+This is critical for debugging clusters that fail to install or fail to upgrade with new operators present.
+
+## Motivation
+
+Debugging information for clusters that succeed in bootstrapping, but fail during installation is missing most
+of the data required to resolve the issue via must-gather.
+
+### Goals
+
+1. allow collection of debugging data for failed installs using normal tools. 
+
+### Non-Goals
+
+1. create a new tool to gather data for failed installs after bootstrapping.
+
+## Proposal
+
+1. `clusteroperator` resources in the payload should be created with the required status conditions (available, progressing,
+   degraded) set to `Unknown`.
+2. `clusteroperator` creation by the CVO needs to honor or update `.status.relatedResources`.  This requires updating
+    status after the creation.
+3.  `clusteroperator` resources in the payload should all be created immediately regardless of where in the payload ordering
+    they are located.  This ensures that they are always present during collection.
```

Comment:
can we define immediately ? also why is this behavior required?

Reply:
does this introduce a new API that operators can assume the these conditions will be pre-set..?

Reply:
> does this introduce a new API that operators can assume the these conditions will be pre-set..?

Not in any reasonable way.  They are controllers.



> can we define immediately ? also why is this behavior required?

As noted in motivation above, we need the metadata to be able to collect from failed installs.  Immediately I would define as searching the entire payload for clusteroperators and creating them before anything else.

Reply:
Ugh, this is going to break a lot of progress notification stuff.  This is hugely ugly and may revert end user stuff.  We will probably have to lie about progress - do a first pass create attempt, but if it fails do not report it at all.

Reply:
I would rather not special-case ordering for a particular type.  We [already provide a way for operators to declare manifest ordering][1], let's just use that instead of complicating it.  Having operators each PR to shift their ClusterOperator manifest to the front of their block is enough to get them all safely up by the time there is any operator-specific content to gather.  If an operator-adding update gets hung up early without pushing the ClusterOperator for the new operator, then there would be no other components associated with that new operator (namespace, deployments, etc.) around to gather.

[1]: https://github.com/openshift/cluster-version-operator/blob/2671e3588883a6e672533e96baae404951df1ef4/docs/dev/operators.md#what-is-the-order-that-resources-get-createdupdated-in